### PR TITLE
OCPBUGS-104597: fix(etcdbackup): correctly classify Azure client-secret credentials from Velero BSL secrets

### PR DESCRIPTION
**Pull request by rutvik23**
## Summary
  
- `HCPEtcdBackup` misclassifies Azure service-principal credentials as WorkloadIdentity when the credential secret originates from a Velero BackupStorageLocation (OADP flow).

-   **Root cause:** `resolveAzureCredentials` scanned the `cloud` dotenv key for
  `AZURE_CLIENT_ID` and `break`-ed immediately on finding it, never reading
  `AZURE_CLIENT_SECRET`. A secret with both keys was classified as WorkloadIdentity instead
  of ClientSecret, causing the reconciler to:
    - Annotate the backup job SA with the Workload Identity(WI) annotation
    - Inject the WI projected token into the pod
    - Omit `--credentials-file` from the etcd-upload container args
    - `etcd-upload` then fell through to `DefaultAzureCredential`, which failed because no
  federated identity is configured for the service principal.

  **Changes:**
  
  - `hypershift-operator/controllers/etcdbackup/credentials.go` — fix credential
    classification in the `cloud` key branch: scan all lines for both `AZURE_CLIENT_ID` and
    `AZURE_CLIENT_SECRET`, then apply priority (both → ClientSecret, ID only →
    WorkloadIdentity, neither → ManagedIdentity).

  - `etcd-upload/azure_uploader.go` — add dotenv fallback in `newClientSecretCredential`:
    after JSON parse fails, scan lines for `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and
    `AZURE_TENANT_ID`. Required because the reconciler mounts the raw `credentials` key from
    the BSL secret, which is dotenv format, not JSON.

  ## Test plan

  - [x] New unit tests in `credentials_test.go` cover: `cloud` key with CLIENT_ID +
    CLIENT_SECRET → `AzureClientSecret`; CLIENT_SECRET before CLIENT_ID (order independence);
    CLIENT_SECRET without CLIENT_ID → `ManagedIdentity`
  - [x] New unit tests in `azure_uploader_test.go` cover: dotenv credentials file → credential
    created successfully; dotenv missing `AZURE_TENANT_ID` → error
  - [x] make tests & all existing tests pass (`go test ./hypershift-operator/controllers/etcdbackup/...
    ./etcd-upload/...`)

## Which issue(s) this PR fixes:
 Fixes: [OCPBUGS-104597](https://redhat.atlassian.net/browse/OCPBUGS-104597)

## Summary by CodeRabbit

* **New Features**
  * Added support for loading Azure client-secret credentials from dotenv-style files.
  * Preserved workload-identity and managed-identity authentication when applicable.
  * Added flexible parsing for whitespace, duplicate values, Windows line endings, and values containing equals signs.
  * Improved credential Secret handling by mounting only the required credentials entry.

* **Bug Fixes**
  * Improved credential detection regardless of environment-variable order.
  * Added clear errors when required Azure credentials are missing.

### CODE REVIEW

File: etcd-upload/azure_uploader.go [hunk 1]
```diff
@@ -130,17 +131,32 @@ func newAzureCredential(ctx context.Context, credentialsFile, authType string) (
...
 func newClientSecretCredential(credentialsFile string) (azcore.TokenCredential, error) {
 	data, err := os.ReadFile(credentialsFile)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read credentials file %q: %w", credentialsFile, err)
 	}
 
 	var creds azureCredentialsFile
-	if err := json.Unmarshal(data, &creds); err != nil {
-		return nil, fmt.Errorf("failed to parse credentials file: %w", err)
+	if jsonErr := json.Unmarshal(data, &creds); jsonErr != nil {
```
Comment (hlipsig) [inline etcd-upload/azure_uploader.go:146]: I know this is in draft, but I thought some ARO perspective might be useful here. As I see it, ARO-RP’s pattern for using azidentity appears applicable here: parse credentials into a typed object at the input boundary, validate them, and pass them explicitly to the matching  azidentity  constructor with centralized cloud-aware options. Could we extract a shared Velero Azure credential parser and use  azureutil.GetAzureCloudConfiguration  rather than duplicating line parsing and discarding  AZURE_CLOUD_NAME ?  azidext  itself is only a compatibility adapter and does not parse credentials.

Comment (rutvik23) [reply etcd-upload/azure_uploader.go:146]: @hlipsig - appreciate your feedback!
- So I've made the changes, specifically refactored the two dotenv parsers this fix introduced into a shared, dependency-free support/azureutil/velerocreds package, wired onto both call sites (resolveAzureCredentials and newClientSecretCredential), so the line-scan logic lives in one place.
- Classification stays presence-based/lenient while the uploader keeps its strict "clientID+secret+tenantID required" validation and error string locally; the parser now also captures AZURE_CLOUD_NAME instead of discarding it (defaulting to AzurePublicCloud).
- Regarding azidext, this repo doesn't seem to have any reference; thus, moving ahead with the parsing which is being done by the new velerocreds package and credentials are passed directly to the azidentity constructor, so there's no adapter in the path. 
- Cloud-aware options via GetAzureCloudConfiguration are intentionally left for a follow-up while keeping the scope of misclassification in OCPBUGS-104597 (and keeping velerocreds dependency-free avoids pulling azureutil into the etcd-upload binary).

Comment (hlipsig) [reply etcd-upload/azure_uploader.go:146]: Awesome. I'll look again.

Comment (hlipsig) [reply etcd-upload/azure_uploader.go:146]: No new feedback. Feel free to request a full review from me when you're ready to take out of draft.

**Comment by rutvik23** [comment]
#### Reproducer : **Method** - Direct credential secret creation (no OADP required) 
```
$ kubectl get hcpetcdbackup -A
NAMESPACE           NAME                  COMPLETED   URL   AGE
clusters-rkhcp-hc   repro-104597-backup   False             5m7s

% kubectl get hcpetcdbackup -A -o yaml
apiVersion: v1
items:
- apiVersion: hypershift.openshift.io/v1beta1
  kind: HCPEtcdBackup
  metadata:
    name: repro-104597-backup
    namespace: clusters-rkhcp-hc
    resourceVersion: "190031"
    uid: 0204b144-554d-445b-bf4b-7904cff638d1
  spec:
    storage:
      azureBlob:
        container: etcd-backups
        credentials:
          name: etcd-backup-test-creds
        keyPrefix: repro-104597
        storageAccount: hcpetcdbackup188411
      storageType: AzureBlob
```
```
% kubectl get job -n hypershift -l app=etcd-backup
NAME                                    STATUS    COMPLETIONS   DURATION   AGE
etcd-backup-repro-104597-backup-ds5p4   Running   0/1           8m7s       8m7s

% kubectl get sa  etcd-backup-job  -n hypershift -o yaml
- apiVersion: v1
  kind: ServiceAccount
  metadata:
    annotations:
      azure.workload.identity/client-id: 7995e17d-ecc2-xyz-xyz-3bb389ad9985 <-----
    creationTimestamp: "2026-08-31T15:20:00Z"
    name: etcd-backup-job
    namespace: hypershift
```
```
$ kubectl get secret etcd-backup-test-creds -n hypershift 
NAME                     TYPE     DATA   AGE
etcd-backup-test-creds   Opaque   1      24m

$ kubectl get pod -n hypershift -l app=etcd-backup -o yaml | grep -A 20 "labels:" | head -10
    labels:
      app: etcd-backup
      azure.workload.identity/use: "true" <---
```
```
$ kubectl logs etcd-backup-repro-104597-test-final-d5lgs-fz5ws
Error: failed to upload snapshot: failed to upload to Azure Blob etcd-backups/repro-104597-final/1788196326.db: DefaultAzureCredential: failed to acquire a token.
Attempted credentials:
    EnvironmentCredential: missing environment variable AZURE_TENANT_ID
    WorkloadIdentityCredential: no client ID specified. Check pod configuration or set ClientID in the options
```
-------
### After fix (Built CPO, HO container image from the fix branch)
#### Check 1: Job completes successfully
```
% cat <<EOF | oc apply -f -
apiVersion: hypershift.openshift.io/v1beta1
kind: HCPEtcdBackup
metadata:
  name: fix-104597-success
  namespace: clusters-rkhcp-hc
spec:
  storage:
    storageType: AzureBlob
    azureBlob:
      container: etcd-backups
      storageAccount: hcpetcdbackup188411
      keyPrefix: fix-104597-success
      credentials:
        name: etcd-backup-test-creds
EOF

% kubectl get pods              
NAME                                         READY   STATUS      RESTARTS       AGE
debug-creds                                  1/1     Running     0              34m
etcd-backup-fix-104597-success-2xtbm-222jt   0/1     Completed   0              11s  <---
external-dns-57fd48f4b9-l578c                1/1     Running     0              134m
operator-d55d5dc84-b9zbn                     1/1     Running     1 (133m ago)   134m
operator-d55d5dc84-x6sdq                     1/1     Running     1 (133m ago)   134m
```
```
% kubectl logs etcd-backup-fix-104597-success-2xtbm-222jt
Defaulted container "upload" out of: upload, fetch-certs (init), snapshot (init)
https://hcpetcdbackup188411.blob.core.windows.net/etcd-backups/fix-104597-success/1788244276.db
```
#### Check 2: NO Workload Identity annotation
```
% kubectl get sa etcd-backup-job -o yaml
apiVersion: v1
imagePullSecrets:
- name: pull-secret
kind: ServiceAccount
metadata:
  creationTimestamp: "2026-09-01T05:34:02Z"
  name: etcd-backup-job
  namespace: hypershift

Also verified both keys exist and contain dotenv content
kubectl get secret etcd-backup-test-creds -n "$HO_NS" -o jsonpath='{.data}' | jq 'keys'
kubectl get secret etcd-backup-test-creds -n "$HO_NS" -o jsonpath='{.data.credentials}' | base64 -d

Expected output: keys `["cloud","credentials"]`, each holding dotenv-formatted key=value pairs (not JSON).
```
#### Check 3: Pod has --credentials-file and --azure-auth-type=client-secret
```
apiVersion: v1
kind: Pod
metadata:
  generateName: etcd-backup-fix-104597-success-2xtbm-
  labels:
    app: etcd-backup
    hypershift.openshift.io/backup-name: fix-104597-success
    hypershift.openshift.io/hcp: rkhcp-hc
    hypershift.openshift.io/hcp-namespace: clusters-rkhcp-hc
    job-name: etcd-backup-fix-104597-success-2xtbm
    topology.kubernetes.io/region: eastus
    topology.kubernetes.io/zone: "0"
  name: etcd-backup-fix-104597-success-2xtbm-222jt
  namespace: hypershift
  ownerReferences:
  - apiVersion: batch/v1
    blockOwnerDeletion: true
    controller: true
    kind: Job
    name: etcd-backup-fix-104597-success-2xtbm
spec:
  containers:
  - command:
    - control-plane-operator
    - etcd-upload
    - --snapshot-path
    - /etc/etcd-backup/snapshot.db
    - --storage-type
    - AzureBlob
    - --azure-container
    - etcd-backups
    - --azure-storage-account
    - hcpetcdbackup188411
    - --key-prefix
    - fix-104597-success
    - --credentials-file
    - /etc/etcd-backup-creds/credentials  <----
    - --azure-auth-type
    - client-secret  <----
    image: quay.io/rkshirsa/hypershift:hcpetcdbkp2026-08-31
```
```
% kubectl get hcpetcdbackup 
NAMESPACE           NAME                 COMPLETED   URL                                                                                               
clusters-rkhcp-hc   fix-104597-success   True        https://hcpetcdbackup188411.blob.core.windows.net/etcd-backups/fix-104597-success/1788244276.db   
```

File: etcd-upload/azure_uploader.go (hunk 1 above)
Comment (rutvik23) [reply etcd-upload/azure_uploader.go:146]: I've opened it for full review, also added reproducer and verification results here: 
https://github.com/openshift/hypershift/pull/9321#issuecomment-5490797845

**Review by hlipsig — APPROVED** [review]
No notes on my side from the Azure interactions.

File: hypershift-operator/controllers/etcdbackup/credentials.go [hunk 2]
```diff
@@ -84,24 +85,27 @@ func resolveAWSCredentials(secret *corev1.Secret) resolvedCredentials {
...
-		var clientID string
-		for line := range strings.SplitSeq(string(cloudData), "\n") {
-			line = strings.TrimSpace(line)
-			if v, ok := strings.CutPrefix(line, "AZURE_CLIENT_ID="); ok {
-				clientID = strings.TrimSpace(v)
-				break
+		// The 'cloud' key is Velero dotenv. Presence of AZURE_CLIENT_SECRET
+		// alongside AZURE_CLIENT_ID marks a service-principal (client-secret)
+		// credential; AZURE_CLIENT_ID alone marks workload identity; neither
+		// marks managed identity.
```
Comment (jparrill) [inline hypershift-operator/controllers/etcdbackup/credentials.go:91]: Hey, the classification fix looks correct — this was definitely the root cause. But I think there's a downstream issue: when this returns `credentialModeAzureClientSecret` from the `"cloud"` key path, `needsCredentialsFile()` returns true, and `buildUploadArgs` passes `--credentials-file /etc/etcd-backup-creds/credentials`. Since the secret's data key is `"cloud"` (not `"credentials"`), the mounted file will be at `/etc/etcd-backup-creds/cloud`, and the uploader will get a file-not-found error.

File: etcd-upload/azure_uploader.go [hunk 3]
```diff
@@ -130,17 +132,28 @@ func newAzureCredential(ctx context.Context, credentialsFile, authType string) (
...
+		// Fallback: parse Velero dotenv format (KEY=value lines). The strict
+		// requirement that all three fields be present stays local to the
+		// uploader; the shared parser only extracts whatever is present.
+		parsed := velerocreds.ParseDotenv(data)
+		creds.ClientID = parsed.ClientID
+		creds.ClientSecret = parsed.ClientSecret
```
Comment (jparrill) [inline etcd-upload/azure_uploader.go:152]: When the dotenv fallback also fails, the error wraps only the JSON error. An operator debugging a Velero BSL secret (which is dotenv, not JSON) would see a confusing JSON unmarshal error. Something like:
```go
return nil, fmt.Errorf("failed to parse credentials file as JSON (%w) or dotenv (missing required fields)", jsonErr)
```
would make troubleshooting much easier.

File: support/azureutil/velerocreds/velerocreds.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// It is intentionally dependency-free (standard library only) so that lean
+// consumers such as the etcd-upload binary do not have to pull in the heavier
+// support/azureutil package (and its controller-runtime / armnetwork deps).
+package velerocreds
+
+import "strings"
+
+// DefaultCloudName is the Azure cloud environment assumed when a Velero
+// credential blob does not specify AZURE_CLOUD_NAME.
+const DefaultCloudName = "AzurePublicCloud"
```
Comment (jparrill) [inline support/azureutil/velerocreds/velerocreds.go:21]: This duplicates `support/config/constants.go:DefaultAzureCloud` which is already `"AzurePublicCloud"`. I get that you wanted to keep this package stdlib-only — but since `support/config` only imports stdlib too, you could use it here without pulling in heavy deps. Worth checking.

**Review by jparrill — COMMENTED** [review]
Minor: the strings `"client-secret"` and `"managed-identity"` in `azureAuthType()` (credentials.go:50,52) are also defined as constants in `etcd-upload/azure_uploader.go` (`AuthTypeClientSecret`, `AuthTypeManagedIdentity`). These two packages form a pipeline — controller picks the mode, uploader consumes it. Having the values hardcoded in two places is a latent bug if one side changes. Would it make sense to put the constants in `velerocreds` (or another shared spot) since you're already introducing that package?

File: support/azureutil/velerocreds/velerocreds.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+// construction requires ClientID, ClientSecret and TenantID).
+type Credentials struct {
+	ClientID     string
+	ClientSecret string
+	TenantID     string
+	// CloudName is the value of AZURE_CLOUD_NAME, defaulting to DefaultCloudName
+	// when the blob omits it. It is captured (rather than discarded) so callers
+	// can make cloud-aware decisions; today the etcd-backup flow only supports
+	// the public cloud.
+	CloudName string
```
Comment (jparrill) [inline support/azureutil/velerocreds/velerocreds.go:35]: Neither caller uses `CloudName` today. Would you be open to leaving it unexported for now? Exporting later is backwards-compatible; unexporting is not. Keeps the API surface minimal until there's a real consumer.

File: hypershift-operator/controllers/etcdbackup/credentials_test.go [hunk 6]
```diff
@@ -177,6 +177,49 @@ func TestResolveAzureCredentials(t *testing.T) {
...
+			ObjectMeta: metav1.ObjectMeta{Name: "azure-creds"},
+			Data: map[string][]byte{
+				"cloud": []byte("AZURE_SUBSCRIPTION_ID=sub-123\nAZURE_TENANT_ID=tenant-456\nAZURE_CLIENT_SECRET=sp-secret\n"),
+			},
+		}
+
+		result := resolveAzureCredentials(secret)
+		g.Expect(result.Mode).To(Equal(credentialModeAzureManagedIdentity))
+		g.Expect(result.ClientID).To(BeEmpty())
+	})
```
Comment (jparrill) [inline hypershift-operator/controllers/etcdbackup/credentials_test.go:221]: nit: Could you add a boundary test for whitespace-only `AZURE_CLIENT_SECRET`? Something like:
```go
"cloud": []byte("AZURE_CLIENT_ID=client-789\nAZURE_CLIENT_SECRET=   \n")
```
After `ParseDotenv` trims, `ClientSecret` should be empty, so this should resolve to WorkloadIdentity, not ClientSecret. It's the exact edge that sits on the boundary of the bug fix.
